File: Data Structures & Algorithms/house-robber/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        memo = {
            0: nums[0],
            1: max(nums[0], nums[1])
        }
        def backtrack(i):
            if i == len(nums):
                return 
            
            memo[i] = max(memo[i-2] + nums[i], memo[i-1])
            backtrack(i+1)

        backtrack(2)
        return memo[len(nums) - 1]


    

File: Data Structures & Algorithms/maximum-subarray/submission-2.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) 
    {
        int n = nums.size();
        if (n == 1)
        {
            return nums[0];
        }
        int res = nums[0];
        int curr = 0;


        for (int i = 0; i < n; i ++)
        {
            if (curr < 0)
            {
                curr = 0;
            }
            curr += nums[i];
            res = max(curr, res);
        }
        return res;

    }
};